legacy/firmware/ethereum.c

```diff
@@ -987,12 +987,27 @@ static void ethereum_typed_hash(const uint8_t domain_separator_hash[32],
   keccak_Final(&ctx, hash);
 }
 
+// When primaryType="EIP712Domain", we ignore message_hash completely
+static void ethereum_domain_only_typed_hash(
+    const uint8_t domain_separator_hash[32], uint8_t hash[32]) {
+  struct SHA3_CTX ctx = {0};
+  sha3_256_Init(&ctx);
+  sha3_Update(&ctx, (const uint8_t *)"\x19\x01", 2);
+  sha3_Update(&ctx, domain_separator_hash, 32);
+  keccak_Final(&ctx, hash);
+}
+
 void ethereum_typed_hash_sign(const EthereumSignTypedHash *msg,
                               const HDNode *node,
                               EthereumTypedDataSignature *resp) {
   uint8_t hash[32] = {0};
-  ethereum_typed_hash(msg->domain_separator_hash.bytes, msg->message_hash.bytes,
-                      hash);
+
+  if (msg->message_hash.size) {
```

Comment:
As uncovered by CI in https://gitlab.com/satoshilabs/trezor/trezor-firmware/-/jobs/1934782475, we cannot use this condition, as current state expects all the hashes to be size 32, and will fail otherwise. The check is at https://github.com/trezor/trezor-firmware/blob/master/legacy/firmware/fsm_msg_ethereum.h#L223-L226

We could theoretically just extend that check with `&& msg->message_hash.size != 0` to allow for this new possibility

Reply:
Done in https://github.com/trezor/trezor-firmware/pull/2036/commits/2d1a145b172ce6e20cce496e9ba2ba05e4863f76 (although I haven't tested it yet locally, so I'll let CI test it). I've made it a `squash! ...` commit so that the commit message will be updated.